### DISCUSSION TITLE
How would you automate generating a `MakeSut(out ...)` method from any constructor?

### DISCUSSION THREAD

**Discussion by szalapski** [devlooped/moq#1447] [Ideas]
Given any constructor of a class I want to test, such as:

```c#
    public MyService(IFooStore store, IBarService service, IModeratelyLongDependencyName dependency3, ...)
    {
       ...
    }
```

Can you think of a good way to **automatically generate a MakeSut method like the following**?  I think this is the most tedious part of writing tests for me.  

```c#
private static MyService MakeSut(
    out Mock<IFooStore> mockStore,
    out Mock<IBarService> mockService,
    out Mock<IModeratelyLongDependencyName> mockDependency3,
    ... )
{ 
  mockStore = new();
  mockService = new();
  mockDependency3 = new();

 return new MyService (mockStore.Object,
     mockStore.Object,
     mockService.Object,
     mockDependency3 .Object,
     ...
);
```

I have found this to be a good pattern, as the out-parameters let me `Setup` what I want and ignore what I don't care about when doing non-strict tests. Having this method with the out parameters lets me instantiate all manner of different SUTs easily like...
```c#
// arrange
MyService sut = MakeSut(out var mockStore,  out _, out _, ...);
mockStore.Setup(s => s.GetFoo(It.IsAny<int>())).ReturnsAsync(new Foo(){...});

// act
sut.DoTheThing();

// assert
mockStore.Verify(...);
```

I've explored doing this using Resharper's Custom Patterns, but it seems too limited to accomplish such a thing. **So how would you go about writing something that could generate the code for such a MakeSut method from any constructor?**  

(Or am I barking up the wrong tree and you know of a better pattern to instantiate SUTs with all dependencies mocked and every mock available for `.Setup`?)

**Comment by kzu** [comment]
Hm... you may be onto something here. I guess a source generator could emit that for you...
